File: src/engine/commands/drinkcommand.cpp
#include "drinkcommand.h"

#include "engine/util.h"


DrinkCommand::DrinkCommand(Player *character, QObject *parent) :
    Command(character, parent) {

    setDescription("Drink an item in the current area or your inventory.\n"
                   "\n"
                   "Example: drink potion");
}

DrinkCommand::~DrinkCommand() {
}

void DrinkCommand::execute(const QString &command) {

    setCommand(command);

    /*QString alias = */takeWord();
    if (!assertWordsLeft("Drink what?")) {
        return;
    }

    takeWord("the");

    GameObjectPtrList allItems = player()->inventory() + currentArea()->items();
    GameObjectPtrList items = takeObjects(allItems);
    if (!requireSome(items, "You don't have that.")) {
        return;
    }

    Item *item = items[0].cast<Item *>();
    if (!item->hasTrigger("ondrink")) {
        send("You cannot drink that.");
        return;
    }

    if (!item->invokeTrigger("ondrink", player())) {
        return;
    }

    item->setDeleted();
}

File: data/commands/drinkcommand.js
function DrinkCommand() {
    Command.call(this);

    this.setDescription("Drink an item in the current room or your inventory.\n" +
                        "\n" +
                        "Example: drink potion");
}

DrinkCommand.prototype = new Command();
DrinkCommand.prototype.constructor = DrinkCommand;

DrinkCommand.prototype.execute = function(player, command) {

    this.prepareExecute(player, command);

    var pool = player.inventory.concat(player.currentRoom.items);
    var item = this.takeObject(pool);
    if (!this.requireSome(item, "Drink what?")) {
        return;
    }

    if (!item.hasTrigger("ondrink")) {
        player.send("You cannot drink that.");
        return;
    }

    if (!item.invokeTrigger("ondrink", player)) {
        return;
    }

    item.setDeleted();
}

CommandRegistry.registerCommand("drink", new DrinkCommand());
